File: 2030-smallest-k-length-subsequence-with-occurrences-of-a-letter/2030-smallest-k-length-subsequence-with-occurrences-of-a-letter.py
class Solution:
    def smallestSubsequence(self, s: str, k: int, letter: str, repetition: int) -> str:
        
        
        strCnt = s.count(letter)
        stackCnt = 0
        stack = []
        
        
        for i, c in enumerate(s) :
            while stack and (( c < stack[-1] and len(s) - i + len(stack) > k and (stack[-1] != letter or strCnt + stackCnt > repetition))                     or  stackCnt + k - len(stack)  < repetition) :
                    stackCnt  -=  1 if stack.pop() == letter else 0
            if len(stack) < k :
                stackCnt += 1 if c == letter else 0
                stack.append(c)
                
            strCnt -= 1 if c == letter else 0
        return "".join(stack)     

File: 2030-smallest-k-length-subsequence-with-occurrences-of-a-letter/2030-smallest-k-length-subsequence-with-occurrences-of-a-letter.js
/**
 * @param {string} s
 * @param {number} k
 * @param {character} letter
 * @param {number} repetition
 * @return {string}
 */
var smallestSubsequence = function(s, k, letter, repetition) {
    
    const last = (arr) => arr[arr.length -1]
    //curr count in stack for the target letter
    //curr count in remaining str for the target curr
    
    // curr remianin length
    let countInStr = s.split("").reduce((a, b) => a + (b ==letter ? 1 : 0), 0)
    let countInStack = 0
    let stack = []
    for(let i =0; i< s.length; i++){
        let currLetter = s[i]
        while(stack.length > 0 && s.length - i + stack.length - 1 >= k   && 
               ((currLetter < last(stack) && (last(stack) != letter || countInStack + countInStr - 1  >= repetition))
                     || k - stack.length < repetition - countInStack))
               
      {
            
            countInStack -= last(stack) == letter ? 1 : 0 
            stack.pop()
        }
            
        if(stack.length < k){
             countInStack = countInStack +  (currLetter === letter ? 1 : 0)
             stack.push(currLetter)
        }
        countInStr -= currLetter == letter ? 1 : 0 


        
    }
    return stack.join("")
    
};